pyg_lib/csrc/sampler/cpu/subgraph_kernel.cpp

```diff
@@ -15,78 +16,7 @@ std::tuple<at::Tensor, at::Tensor, c10::optional<at::Tensor>> subgraph_kernel(
-  TORCH_CHECK(nodes.is_cpu(), "'nodes' must be a CPU tensor");
-
-  const auto num_nodes = rowptr.size(0) - 1;
-  const auto out_rowptr = rowptr.new_empty({nodes.size(0) + 1});
-  at::Tensor out_col;
-  c10::optional<at::Tensor> out_edge_id = c10::nullopt;
-
-  AT_DISPATCH_INTEGRAL_TYPES(nodes.scalar_type(), "subgraph_kernel", [&] {
-    auto mapper = pyg::sampler::Mapper<scalar_t>(num_nodes, nodes.size(0));
-    mapper.fill(nodes);
-
-    const auto rowptr_data = rowptr.data_ptr<scalar_t>();
-    const auto col_data = col.data_ptr<scalar_t>();
-    const auto nodes_data = nodes.data_ptr<scalar_t>();
-
-    // We first iterate over all nodes and collect information about the number
-    // of edges in the induced subgraph.
-    const auto deg = rowptr.new_empty({nodes.size(0)});
-    auto deg_data = deg.data_ptr<scalar_t>();
-    auto grain_size = at::internal::GRAIN_SIZE;
-    at::parallel_for(0, nodes.size(0), grain_size, [&](int64_t _s, int64_t _e) {
-      for (size_t i = _s; i < _e; ++i) {
-        const auto v = nodes_data[i];
-        // Iterate over all neighbors and check if they are part of `nodes`:
-        scalar_t d = 0;
-        for (size_t j = rowptr_data[v]; j < rowptr_data[v + 1]; ++j) {
-          if (mapper.exists(col_data[j]))
-            d++;
-        }
-        deg_data[i] = d;
-      }
-    });
-
-    auto out_rowptr_data = out_rowptr.data_ptr<scalar_t>();
-    out_rowptr_data[0] = 0;
-    auto tmp = out_rowptr.narrow(0, 1, nodes.size(0));
-    at::cumsum_out(tmp, deg, /*dim=*/0);
-
-    out_col = col.new_empty({out_rowptr_data[nodes.size(0)]});
-    auto out_col_data = out_col.data_ptr<scalar_t>();
-    scalar_t* out_edge_id_data;
-    if (return_edge_id) {
-      out_edge_id = col.new_empty({out_rowptr_data[nodes.size(0)]});
-      out_edge_id_data = out_edge_id.value().data_ptr<scalar_t>();
-    }
-
-    // Customize `grain_size` based on the work each thread does (it will need
-    // to find `col.size(0) / nodes.size(0)` neighbors on average).
-    // TODO Benchmark this customization
-    grain_size = std::max<int64_t>(out_col.size(0) / nodes.size(0), 1);
-    grain_size = at::internal::GRAIN_SIZE / grain_size;
-    at::parallel_for(0, nodes.size(0), grain_size, [&](int64_t _s, int64_t _e) {
-      for (scalar_t i = _s; i < _e; ++i) {
-        const auto v = nodes_data[i];
-        // Iterate over all neighbors and check if they are part of `nodes`:
-        scalar_t offset = out_rowptr_data[i];
-        for (scalar_t j = rowptr_data[v]; j < rowptr_data[v + 1]; ++j) {
-          const auto w = mapper.map(col_data[j]);
-          if (w >= 0) {
-            out_col_data[offset] = w;
-            if (return_edge_id)
-              out_edge_id_data[offset] = j;
-            offset++;
-          }
-        }
-      }
-    });
-  });
-
-  return std::make_tuple(out_rowptr, out_col, out_edge_id);
+  return subgraph_bipartite(rowptr, col, nodes, nodes, return_edge_id);
```

Comment:
The code structure looks a little weird to me because `csrc/sampler/cpu/subgraph_kernel` exists for register `TORCH_LIBRARY_IMPL` and it is using a general implementation in `csr/sampler/subgraph.cpp`. How about reorganize the code like this:
```
csr
  - ops
    # all ops expose for pytorch.
    - sampler
  # all general graph operation.
  - sampler
```
We don't need to refactor the code structure now. But want to hear your opinion.

Reply:
nvm, seems subgraph.cpp also defines library. Why not merge them together since sampler/subgraph.cpp also runs on cpu only.

Reply:
We could follow the style in other pyg repos: put CPU/GPU specific impl in separate folders and provide common interface in a higher directory.